--- ../../../../../../../../update/IdpReviewProfileAuthenticatorFactory.java.24.0.3.orig	2024-04-30 10:58:50.496172222 +0300
+++ EvakaReviewProfileAuthenticatorFactory.java	2023-08-08 15:10:57.034961623 +0300
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.keycloak.authentication.authenticators.broker;
+package com.espoo.keycloak.authenticator;
 
 import org.keycloak.Config;
 import org.keycloak.authentication.Authenticator;
@@ -32,11 +32,13 @@
 
 /**
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ * Modified from keycloak/services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpReviewProfileAuthenticatorFactory.java
+
  */
-public class IdpReviewProfileAuthenticatorFactory implements AuthenticatorFactory {
+public class EvakaReviewProfileAuthenticatorFactory implements AuthenticatorFactory {
 
-    public static final String PROVIDER_ID = "idp-review-profile";
-    static IdpReviewProfileAuthenticator SINGLETON = new IdpReviewProfileAuthenticator();
+    public static final String PROVIDER_ID = "evaka-review-profile";
+    static EvakaReviewProfileAuthenticator SINGLETON = new EvakaReviewProfileAuthenticator();
 
     public static final String UPDATE_PROFILE_ON_FIRST_LOGIN = "update.profile.on.first.login";
 
@@ -82,7 +84,7 @@
 
     @Override
     public String getDisplayType() {
-        return "Review Profile";
+        return "Evaka Review Profile";
     }
 
     @Override
